src/Microsoft.Health.Api/Features/ByteCounter/ByteCountingStream.cs

```diff
@@ -0,0 +1,135 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+
+namespace Microsoft.Health.Api.Features.ByteCounter
+{
+    /// <summary>
+    /// A stream to wrap an underlying stream and counts the number of bytes passed while operating with it.
+    /// </summary>
+    public class ByteCountingStream : Stream
+    {
+        private readonly Stream _stream;
+        private long _writtenByteCount;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ByteCountingStream"/> class.
+        /// </summary>
+        /// <param name="stream">An underlying stream.</param>
+        public ByteCountingStream(Stream stream)
+        {
+            EnsureArg.IsNotNull(stream, nameof(stream));
+
+            _stream = stream;
+        }
+
+        /// <inheritdoc />
+        public override bool CanRead => _stream.CanRead;
+
+        /// <inheritdoc />
+        public override bool CanSeek => _stream.CanSeek;
+
+        /// <inheritdoc />
+        public override bool CanWrite => _stream.CanWrite;
+
+        /// <inheritdoc />
+        public override long Length => _stream.Length;
+
+        /// <inheritdoc />
+        public override long Position
+        {
+            get { return _stream.Position; }
```

Comment:
nit: can still use expressions here:
```suggestion
            get => _stream.Position;
            set => _stream.Position = value;
```